### PR TITLE
[WIP][Flang][MLIR][OpenMP][Driver] OpenMP RTL flag attribute and lowering and omp.is_device attribute

### PR DESCRIPTION
**Pull request by agozillon**
This PR builds on the following:

- Sergio's work in: https://github.com/ROCm-Developer-Tools/llvm-project/pull/197 (although it excludes the triple changes as it caused some compilation issues without the triple patch too)
- The previous closed omp.is_device PR: https://github.com/ROCm-Developer-Tools/llvm-project/pull/201

It adds handling and lowering to LLVM-IR of the flags:
- fopenmp-target-debug
- fopenmp-assume-threads-oversubscription
- fopenmp-assume-teams-oversubscription
- fopenmp-assume-no-nested-parallelism
- fopenmp-assume-no-thread-state

Which are emitted to LLVM-IR as globals when device offloading OpenMP in Clang, in the case of Flang we lower it from the frontend to an MLIR attribute that is applied to the module, so that it can be carried down to the OpenMPToLLVMIRTranslation step to be lowered to LLVM-IR as globals.

It adds further handling of the omp.is_device attribute to the bbc tool, so that it can be used to test lowering of Fortran to OpenMP FIR device IR (but it currently doesn't support the RTL flags, as the tool seems intended to be a lightweight Fortran -> FIR test tool). 

It adds full module emission to the bbc tool rather than removing the module on emission of FIR IR, this keeps attributes that have been applied in the bbc lowering process. The issue is that things like omp.is_device or any other attributes will be discarded on emission from the tool. This change does not appear to break any existing tests, but it is something that perhaps needs discussion with the tool author when this is submitted upstream. It seems like a lot of Flang's tooling and testing infrastructure relies on the fact that MLIR tools will implicitly wrap things in a module, this doesn't seem ideal imo.

The addition of module lowering (OMP/LLVM Dialect -> LLVM-IR) infrastructure to the LLVM Translation Interface, this is required for lowering/handling of specialised attributes on modules. But it also enables specialisation of modules for a dialect. There may be an alternative method of lowering the RTL Flags attribute that I'm unaware of, so please do point out if there is a better way to do this!

A lot of tests are still required to be added, but these will be added as I fracture this into smaller phabricator patches as that seems like it'll need to be done if we wish to go ahead with upstreaming this!

### CODE REVIEW

File: flang/include/flang/Frontend/TargetOptions.h [hunk 1]
```diff
@@ -29,6 +29,10 @@ class TargetOptions {
   /// The name of the target triple to compile for.
   std::string triple;
 
+  /// When compiling for the device side, contains the triple used to compile
+  /// for the host.
+  std::string hostTriple;
```
[the remark this answers is not in the document]
Comment (skatrak) [reply flang/include/flang/Frontend/TargetOptions.h:34]: Right, if the community likes this approach of adding dialect attributes to the builtin module for this kind of information, I would expect the host triple to be added in the same way.

In my opinion, we would first need to get the Phabricator patch you refer to approved, as well as adding #194, together with the `hostTriple`-related code from #197 that you had to remove from this patch. Then, we could store the `aux-triple` information in an `omp.host_triple` or similar attribute in the top-level MLIR module. Hopefully that makes sense.

[the remark this answers is not in the document]
Comment (agozillon) [reply flang/include/flang/Frontend/TargetOptions.h:34]: That does indeed make sense! Hopefully the patch is accepted so we can upstream these flag additions.

**Comment by agozillon** [comment]
Phabricator patches related to this work:

https://reviews.llvm.org/D144864 - omp.is_device attribute and the -openmp-is-device flag, the most important

https://reviews.llvm.org/D144896 - The RTLModuleFlagsAttr for the OpenMP dialect which holds the Flag information from the frontend to be lowered to LLVM-IR by the OpenMP/LLVM dialect -> LLVM-IR step

https://reviews.llvm.org/D144883 - Adding the convertModuleOperation to the MLIR -> LLVM translation interface, important, but this may not be the way to do it from the communities perspective, which would mean going back to the drawing board for the lowering of the RTL flags 

https://reviews.llvm.org/D144869 - emitting the full module and not just the body from the bbc tool, good to seek clarity on this but this work doesn't hinge on it

Still another one or two that will follow these if they are successful (lowering of the RTL Flags attribute and the creation of it based on the newly added to Flang flags), I've left them out for now as they are quite dependent on the other patches, so I thought I'd keep it simple for now for myself and the reviewers.

**Comment by agozillon** [comment]
This patch is now fully upstreamed, so I am closing it!

**Pull request closed without merging**